## Architect review — CI failure needs fixing before merge

CI Unit Tests job fails on `run-removed.test.ts`:

```
Error [ERR_MODULE_NOT_FOUND]: Cannot find module '.../dist/commands/porch/index.js'
imported from .../bin/porch.js
    at finalizeResolution (node:internal/modules/esm/resolve:283:11)
```

The regression test spawns `node bin/porch.js run some-id` as a subprocess. `bin/porch.js` imports from `dist/`, but the unit test job doesn't build first (only `test:e2e` does `pnpm build && vitest run`). Locally it passes because dist is already built — CI starts clean.

**Fix**: rewrite the test to call `cli()` directly instead of via subprocess. The bugfix-711 PR (#714) does this for `porch --version`:

```ts
import { cli } from '../index.js';

it('exits 1 with a migration message pointing to porch next', async () => {
  const stderr: string[] = [];
  const spy = vi.spyOn(console, 'error').mockImplementation((...a) => {
    stderr.push(a.map(String).join(' '));
  });
  // Note: cli() throws or process.exits — wrap accordingly
  await expect(cli(['run', 'some-id'])).rejects.toThrow(); // or check process.exit mock
  expect(stderr.join('\n')).toContain("'porch run' has been removed");
  expect(stderr.join('\n')).toContain('porch next');
  spy.mockRestore();
});
```

(Adjust to the actual exit mechanism — looking at line 929 of `index.ts`, it logs to `console.error` and presumably `process.exit(1)`s. You may need to mock `process.exit` too.)

The rest of the PR (4 `describe.skip`s with clear comments + the migration message regression test) is correctly scoped. Please fix the test approach and CI should go green.

— Architect

--- a/codev/projects/bugfix-710-e2e-12-porch-single-phase-happ/status.yaml
+++ b/codev/projects/bugfix-710-e2e-12-porch-single-phase-happ/status.yaml
@@ -1,7 +1,7 @@
 id: bugfix-710
 title: e2e-12-porch-single-phase-happ
 protocol: bugfix
-phase: investigate
+phase: fix
 plan_phases: []
 current_plan_phase: null
 gates: {}
@@ -9,7 +9,7 @@ iteration: 1
 build_complete: false
 history: []
 started_at: '2026-04-30T19:33:52.093Z'
-updated_at: '2026-04-30T23:05:42.271Z'
+updated_at: '2026-04-30T23:05:47.776Z'
 pr_history:
   - phase: investigate
     pr_number: 715

--- a/codev/projects/bugfix-710-e2e-12-porch-single-phase-happ/status.yaml
+++ b/codev/projects/bugfix-710-e2e-12-porch-single-phase-happ/status.yaml
@@ -9,4 +9,9 @@ iteration: 1
 build_complete: false
 history: []
 started_at: '2026-04-30T19:33:52.093Z'
-updated_at: '2026-04-30T19:33:52.094Z'
+updated_at: '2026-04-30T23:05:42.271Z'
+pr_history:
+  - phase: investigate
+    pr_number: 715
+    branch: builder/bugfix-710
+    created_at: '2026-04-30T23:05:42.270Z'

--- a/packages/codev/src/commands/porch/__tests__/run-removed.test.ts
+++ b/packages/codev/src/commands/porch/__tests__/run-removed.test.ts
@@ -6,31 +6,55 @@
  * test pins the migration message so the deprecation remains discoverable
  * for anyone (including the E2E test runner) that still invokes the old
  * command.
+ *
+ * Calls `cli()` directly rather than spawning `bin/porch.js` as a
+ * subprocess: unit tests run before `pnpm build`, so `dist/` may be
+ * absent in CI.
  */
 
-import { describe, it, expect } from 'vitest';
-import { spawnSync } from 'node:child_process';
-import * as path from 'node:path';
+import { describe, it, expect, vi, afterEach } from 'vitest';
+import { cli } from '../index.js';
 
-const PORCH_BIN = path.resolve(__dirname, '..', '..', '..', '..', 'bin', 'porch.js');
+function captureRun(args: string[]): { stderr: string; exitCode: number } {
+  const stderr: string[] = [];
+  let exitCode = 0;
+
+  const errSpy = vi.spyOn(console, 'error').mockImplementation((...a: unknown[]) => {
+    stderr.push(a.map(String).join(' '));
+  });
+  const exitSpy = vi.spyOn(process, 'exit').mockImplementation(((code?: number) => {
+    exitCode = code ?? 0;
+    throw new Error('__process_exit__');
+  }) as never);
+
+  try {
+    // cli() is async and the 'run' branch calls process.exit synchronously
+    // inside the switch. Our mock throws to unwind the stack so the test
+    // doesn't hang waiting on the promise.
+    cli(args).catch(() => { /* swallowed: __process_exit__ */ });
+  } finally {
+    errSpy.mockRestore();
+    exitSpy.mockRestore();
+  }
+
+  return { stderr: stderr.join('\n'), exitCode };
+}
 
 describe('porch run (removed)', () => {
-  it('exits 1 with a migration message pointing to porch next', () => {
-    const result = spawnSync('node', [PORCH_BIN, 'run', 'some-id'], {
-      encoding: 'utf-8',
-    });
+  afterEach(() => {
+    vi.restoreAllMocks();
+  });
 
-    expect(result.status).toBe(1);
-    expect(result.stderr).toContain("'porch run' has been removed");
-    expect(result.stderr).toContain("porch next");
+  it('exits 1 with a migration message pointing to porch next', () => {
+    const { stderr, exitCode } = captureRun(['run', 'some-id']);
+    expect(exitCode).toBe(1);
+    expect(stderr).toContain("'porch run' has been removed");
+    expect(stderr).toContain('porch next');
   });
 
   it('exits 1 even when extra flags are passed (e.g. --single-phase)', () => {
-    const result = spawnSync('node', [PORCH_BIN, 'run', 'some-id', '--single-phase'], {
-      encoding: 'utf-8',
-    });
-
-    expect(result.status).toBe(1);
-    expect(result.stderr).toContain("'porch run' has been removed");
+    const { stderr, exitCode } = captureRun(['run', 'some-id', '--single-phase']);
+    expect(exitCode).toBe(1);
+    expect(stderr).toContain("'porch run' has been removed");
   });
 });

